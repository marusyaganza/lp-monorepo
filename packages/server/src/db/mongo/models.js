const { formatIds, formatFilter } = require('../../utils/dbUtils');
const { ObjectId } = require('mongodb');

function createModel(db, table) {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result = await db.collection(table).find(formatFilter(filter));
    const data = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields) {
    const data = { ...fields };
    delete data.id;
    delete data.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    return result?.result;
  }

  async function deleteOne(filter) {
    console.log('filter', filter);
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    return result?.result;
  }

  async function createMany(entries) {
    const formattedEntries = entries.map(entry => ({
      ...entry,
      createdAt: Date.now()
    }));
    const result = await db.collection(table).insertMany(formattedEntries);
    return result.ops.map(formatIds);
  }

  return {
    findOne,
    findMany,
    createOne,
    createMany,
    updateOne,
    deleteOne
  };
}

module.exports = createModel;
