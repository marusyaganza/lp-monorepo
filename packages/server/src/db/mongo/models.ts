import { formatIds, formatFilter } from '../../utils/dbUtils';

import { Table, ModelType, DbData } from '..';

const { ObjectId } = require('mongodb');

function createModel<T extends DbData>(db: any, table: Table): ModelType<T> {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result: T[] = await db.collection(table).find(formatFilter(filter));
    const data: T[] = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields: Partial<T>) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields: Partial<T>) {
    const data = { ...fields };
    delete data?.id;
    delete data?.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    const isOk = result?.result?.ok ? true : false;
    const modifiedCount = result?.result?.nModified;
    return { ok: isOk, modifiedCount };
  }

  async function deleteOne(filter: Partial<T>) {
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    const isOk = result?.result?.ok ? true : false;
    return { ok: isOk };
  }

  return {
    findOne,
    findMany,
    createOne,
    updateOne,
    deleteOne
  };
}

export default createModel;
