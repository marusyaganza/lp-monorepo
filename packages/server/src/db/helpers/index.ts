export * from './sortingHelpers/sortingHelpers';
export * from './general/general';
