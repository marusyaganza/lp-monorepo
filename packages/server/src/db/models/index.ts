import { UserModel, UserModelType } from './User';
import { WordModel, WordModelType } from './Word';
import { WordTagModel, WordTagModelType } from './WordTag';

export interface ModelsType {
  User: UserModelType;
  Word: WordModelType;
  WordTag: WordTagModelType;
}

export const models: ModelsType = {
  User: UserModel,
  Word: WordModel,
  WordTag: WordTagModel
};
