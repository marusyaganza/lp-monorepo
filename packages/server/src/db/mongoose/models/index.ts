import { UserModel } from './User';
import { WordModel } from './Word';

export const models = {
  User: UserModel,
  Word: WordModel
};
