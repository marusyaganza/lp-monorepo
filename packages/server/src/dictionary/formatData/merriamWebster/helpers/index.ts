export * from './general/general';
export * from './resource/resource';
export * from './string/string';
export * from './def/def';
