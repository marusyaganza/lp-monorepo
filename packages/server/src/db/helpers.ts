import { Document, Types } from 'mongoose';
import {
  Maybe,
  Language,
  Game,
  SortBy,
  SortWordsBy
} from '../generated/graphql';

type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

interface DbData extends ModelDataType {
  _id?: string;
}

type DocumentType<T> =
  | (Document<unknown, unknown, T> &
      T & {
        _id: Types.ObjectId;
      })
  | null;

export interface WordsFilterType {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder?: boolean;
  timesToLearn?: number | null;
  gameType?: Game;
  user?: string;
  tags?: string[];
  searchQuery?: string;
}

export interface WordsWithPaginationFilter extends WordsFilterType {
  page?: number;
  limit?: number;
}

export function formatData<T>(data: DocumentType<T>): T | null {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter?.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
const NO_TAGS_ID = '000000000000000000000000';

export function getWordsFilters(filter: WordsFilterType) {
  const {
    user,
    language = Language.English,
    sortBy,
    isReverseOrder,
    gameType,
    tags,
    searchQuery,
    timesToLearn = 5
  } = filter;

  // filter for a specific game
  let gameFilter;
  if (gameType === Game.Conjugation) {
    gameFilter = { conjugation: { $ne: null } };
  }
  if (gameType === Game.Audio) {
    gameFilter = { audioUrl: { $ne: '' } };
  }

  // filter based on tags
  let tagsFilters: { tags: string }[] = [];
  if (tags?.length) {
    tagsFilters = tags
      ?.filter(tag => tag !== NO_TAGS_ID)
      .map(tag => {
        return { tags: tag };
      });
  }

  let noTagsFilter;
  if (tags?.includes(NO_TAGS_ID)) {
    noTagsFilter = { $or: [{ tags: [] }, { tags: null }] };
  }

  let searchFilter;
  if (searchQuery) {
    const searchRegexp = { $regex: new RegExp(searchQuery, 'i') };
    searchFilter = {
      $or: [
        { name: searchRegexp },
        { 'defs.def': searchRegexp },
        { alternativeSpelling: searchRegexp },
        { stems: searchRegexp }
      ]
    };
  }

  // filter based on learning status
  let learningStatusFilters: Record<string, unknown>[] = [];
  if (gameType && sortBy !== SortBy.MemoryRefresher) {
    learningStatusFilters = [
      { isLearned: { $ne: true } },
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
          { [`statistics.${gameType}.successRate`]: null }
        ]
      }
    ];
  }
  if (sortBy === SortBy.MemoryRefresher) {
    learningStatusFilters = [
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
          { isLearned: true }
        ]
      }
    ];
  }

  //sorting
  let orderNum = isReverseOrder ? -1 : 1;

  // words with the highest number of errors should go first
  if (sortBy === SortBy.ErrorCount) {
    orderNum = -1 * orderNum;
  }

  let sort: Record<string, number> = { $natural: orderNum };

  if (sortBy) {
    let propName: string = sortBy;
    if (gameType) {
      propName = `statistics.${gameType}.${sortBy}`;
    }

    if (sortBy !== SortBy.MemoryRefresher) {
      sort = { [propName]: orderNum };
    }
  }

  const filtersArray = [
    gameFilter,
    noTagsFilter,
    searchFilter,
    ...tagsFilters,
    ...learningStatusFilters
  ].filter(Boolean);

  const filters: Record<string, unknown> = {
    user,
    language
  };

  if (filtersArray?.length) {
    filters.$and = filtersArray;
  }

  return {
    sort,
    filters
  };
}
