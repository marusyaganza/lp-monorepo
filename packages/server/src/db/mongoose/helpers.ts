import { Types, Document } from 'mongoose';
import { User, Word } from '../../generated/graphql';
import { DbData } from 'db';

type DocumentType =
  | (Document<unknown, Record<string, any>, User | Word> &
      Omit<
        | User
        | (Word & {
            _id: Types.ObjectId;
          }),
        never
      >)
  | null;

export function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
