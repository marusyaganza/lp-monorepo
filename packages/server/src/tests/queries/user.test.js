import { createTestServer, getErrorMessageFromGQL } from '../helpers';
import { models } from '../mocks/models';
import { user } from '../mocks/data';
import { userQueries } from './queries';

describe('user queries', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });
  test('user if user exist', async () => {
    const { query } = createTestServer({
      user,
      models
    });

    const res = await query({ query: userQueries.userQuery });
    expect(res).toMatchSnapshot();
  });

  test('user if user is undefined', async () => {
    const { query } = createTestServer({
      models: {
        ...models,
        User: {
          findOne: jest.fn()
        }
      }
    });

    const res = await query({ query: userQueries.userQuery });
    expect(getErrorMessageFromGQL(res)).toEqual('user is not found');
  });

  test('user if user is not found', async () => {
    const { query } = createTestServer({
      user,
      models: {
        ...models,
        User: {
          findOne: jest.fn()
        }
      }
    });

    const res = await query({ query: userQueries.userQuery });
    expect(getErrorMessageFromGQL(res)).toEqual('user is not found');
  });
});
