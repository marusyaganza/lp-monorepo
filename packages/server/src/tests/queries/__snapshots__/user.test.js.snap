// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user queries user if user exist 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "user": {
          "email": "member@member.com",
          "firstName": "User",
          "lastName": "Test",
          "role": "MEMBER",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
